src/NumericalAlgorithms/Spectral/SwshCollocation.hpp

```diff
@@ -0,0 +1,247 @@
+    return CollocationConstIterator{make_not_null(this), size()};
+  }
+  Collocation<Representation>::CollocationConstIterator cend() const noexcept {
+    return end();
+  }
+  // @}
+
+ private:
+  // an extra pointer layer is required for the peculiar way that libsharp
+  // constructs these values.
+  size_t l_max_ = 0;
+  std::unique_ptr<sharp_geom_info*, detail::DestroySharpGeometry> geom_info_;
+};
+
+namespace detail {
+// We use a `std::index_sequence` to loop over the function lambdas
+// for the cases we want to evaluate in order to only compute and store
+// a specific `l_max` that is requested, not all values up to
+// `swsh_collocation_maximum_l_max`. This cannot be done in the more traditional
+// way of either having a `std::vector` or `std::undordered_map` that is
+// `static` because only the construction phase is guaranteed to be thread-safe,
+// not the access. This would mean were we to use a `std::vector` or
+// `std::unordered_map` we would need to compute all values up to
+// `swsh_collocation_maximum_l_max` immediately. By doing the loop over `Is` we
+// are able to generate `sizeof...(Is)` different static objects and thus each
+// one can be constructed only when it is needed and in a thread-safe manner.
+
+template <ComplexRepresentation Representation, size_t... Is>
+SPECTRE_ALWAYS_INLINE const Collocation<Representation>&
+dispatch_to_precomputed_static_collocation_impl(
+    size_t index, std::index_sequence<Is...> /*meta*/) noexcept {
+  if (UNLIKELY(index > swsh_collocation_maximum_l_max)) {
+    ERROR(
+        "The provided l_max is not below the maximum l_max to cache. Either "
+        "construct the Collocation manually, or consider (with caution) "
+        "increasing `swsh_collocation_maximum_l_max`.");
+  }
+  const Collocation<Representation>* collocation = nullptr;
+  auto get_collocation_if_match = [&collocation, &index ](auto i) noexcept {
+    if (UNLIKELY(decltype(i)::value == index)) {
+      static const Collocation<Representation> precomputed_collocation =
+          Collocation<Representation>{decltype(i)::value};
+      collocation = &precomputed_collocation;
+    }
+    return 0;
+  };
+  expand_pack(
+      get_collocation_if_match(std::integral_constant<size_t, Is>{})...);
+  return *collocation;
+}
+}  // namespace detail
+
+/// \brief precomputation function for those collocation grids that are
+/// requested
+///
+/// \details keeps a compile-time structure which acts as a thread-safe lookup
+/// table for all l_max values that have been requested so far during execution,
+/// so that the libsharp generation need not be re-run. If it has been
+/// generated, it's returned by reference. Otherwise, the new grid is generated
+/// and put in the lookup table before it is returned by reference.
+template <ComplexRepresentation Representation>
+const Collocation<Representation>& precomputed_spherical_harmonic_collocation(
```

Comment:
This definition could probably go in the cpp file and then be instantiated explicitly for the two cases. Then all the detail stuff can also go into the cpp file and we only have to compile it once :)